RFC-0022-Model-Partitioning-in-Pipeline-Parallelism.md

```diff
@@ -0,0 +1,255 @@
+
+***PipeDream***
+More on asynchronous training pipeline planning, like parameter server approaches, the algorithm itself also uses Dynamic Programming, but the states is highly correlated with the communication and synchronization costs, which is not a good choice for us (unless we consider async training in the future). 
+
+***DAPPLE***
+
+DAPPLE combines DDP with pipeline parallelism to form a bigger search space and use a more efficient schedule if possible. Their partition approach is a DP-based algorithm, it first tries to find the “pivotal” stage, then optimize the overall latency,  the “overall latency” optimization here tries to reduce the bubbles in the pipeline as small as possible.
+
+
+
+## Proposing: Automated Model Partitioning in PyTorch Pipeline Parallelism
+
+Given the existing work and their limitations, we introduce a set of APIs that’s flexible enough for future improvements and intuitive to use. We expose a single API `create_balanced_partition`, to take the model and do the partition under the hood.
+
+```
+def create_balanced_partition(model: nn.Module,
+                              devices: List[RemoteDevice],           
+                              *sample_input) -> List[RemoteModule]:
+ #device = torch.device("cuda")):
+ # if model itself if larger than a single cuda device memory, 
+ # should we allow the user to profile the model on cpu?
+ # Decision: probably no, as the characteristics are different
+
+ # Step 1
+ # fx profiler to collect statistics based on a single run of
+ # the sample_input
+ interp = ProfilingInterpreter(model)
+ interp.run(sample_input)
+ 
+ # Step2
+ # Partition algorithm to calculate the least 
+ # cost partition assignment
+ num_partitions = len(devices)
+ partitions = _partition_by_cost(interp, num_partitions)
+
+ # Step3
+ # Use the results from Step 2 to assign each part of submodule to the device
+ # returned a fx-based model with partition results applied (assigned to devices already)
+ return partitioned_model
+```
+
+
+Note that this will be the only API we expose to the user, it returns a partitioned model which already been transferred to the corresponding devices. This allow us to iterate on the underlying implementation and experiment more efficient partition algorithms in the future.
+
+## FX compatibility
+
+Pipeline parallelism auto partition capability needs more advanced knowledge in order to accurately divide a model into a balanced partition set.  In order to achieve auto partitioning with the most balanced approach, we need to get the model execution graph and try to split the model base on the graph. Using torch.fx can give us a helpful graph representation in order for us to do more advanced partition with extensive analysis with tracing and profiling. 
```

Comment:
`torch.fx` gives a representation of a function no? Typically in case of a module we'd trace `forward`. However a lot of model have "secondary" forward function, typically `infer` which might be slightly different from the forward function. How does optimizing on the `forward` function, generalises to the `infer` function.